system/lib/pthread/emscripten_futex_wait.c

```diff
@@ -127,12 +128,28 @@ int emscripten_futex_wait(volatile void *addr, uint32_t val, double max_wait_ms)
     return ret;
   }
 
+  // If dynamic linking is supported then we busy wait so that we can handle
+  // loading of new code when another thread calls dlopen.
   // -1 (or any negative number) means wait indefinitely.
   int64_t max_wait_ns = -1;
   if (max_wait_ms != INFINITY) {
     max_wait_ns = (int64_t)(max_wait_ms*1000*1000);
   }
+#ifdef __pic__
+#define MAX_WAIT (1000*1000*100)
+  // If dynamic linking is supported then we only wait for MAX_WAIT nanoseconds
+  // at a time so we can response to dlopen calls that might happen on other
+  // threads.
+  int64_t total = 0;
+  while (max_wait_ns == -1 || total < max_wait_ns) {
+    ret = __builtin_wasm_memory_atomic_wait32((int*)addr, val, MAX_WAIT);
+    if (ret != 2) break;
+    _emscripten_yield();
+    total += MAX_WAIT;
+  }
+#else
   ret = __builtin_wasm_memory_atomic_wait32((int*)addr, val, max_wait_ns);
+#endif
```

Comment:
Changing the fundamental way futexes work based on dlopen being set or not seems confusing to me. Is this not a risk for odd bugs?

Reply:
I mean its not risk free.  I think the main issue here is that it prevents threads from sleeping for more than 100ms at a time.   Note that this is already true when threads call `usleep` or `sleep` or `nano_sleep`:
https://github.com/emscripten-core/emscripten/blob/43074846f216abad2a1c35b5a631f2a9481d9e07/system/lib/pthread/library_pthread.c#L92-L93

I'm inclined to say that the code of this solution is worth it since that alternative is to have threads that in an unknown state regarding what code they have loaded.  That seems like a much more likely cause of strange bugs.

Reply:
I agree that threads in an unsynchronized state are worrying. Still, this is a fundamental change - can we do something more specific? That is, can we not do a generic yield here, which might cause other things to happen, but instead just do a dlopen-specific operation?

Reply:
Sure can do the for now.  As it happens all the `_emscripten_yeidl` currently does on background threads is call `emscripten_process_dlopen_queue`.. so its the same thing today.